File: capture.cpp

#include <opencv2/opencv.hpp>

#include <iostream>

using namespace cv;
using namespace std;

std::string get_tegra_pipeline(int width, int height, int fps, int sensor_id) {
    return "nvcamerasrc sensor-id=(int)" + std::to_string(sensor_id) + "! video/x-raw(memory:NVMM), width=(int)" + std::to_string(width) + ", height=(int)" +
           std::to_string(height) + ", format=(string)I420, framerate=(fraction)" + std::to_string(fps) +
           "/1 ! nvvidconv flip-method=0 ! video/x-raw, format=(string)BGRx ! videoconvert ! video/x-raw, format=(string)BGR ! appsink";
}

int main(int argc, char* arg[])
{
    int id = 0;
    if (argc >= 2)
    {
        std::string num_string = arg[1];
        id = std::stoi(num_string);

        std::cout<<"sensor id = "<<id<<std::endl;
    }
    #if 1
    int WIDTH = 1920;
    int HEIGHT = 1080;
    int FPS = 30;

    // Define the gstream pipeline
    std::string pipeline = get_tegra_pipeline(WIDTH, HEIGHT, FPS, id);
    std::cout << "Using pipeline: \n\t" << pipeline << "\n";

    // Create OpenCV capture object, ensure it works.
    cv::VideoCapture capture(id);
    #else
    VideoCapture capture(id);
    #endif
    Mat frame;
    if (capture.isOpened()) // 判断视频是否成功打开

    {
        while (capture.grab())
        {
            capture >> frame;

            std::cout << "input image size = [" << frame.cols << ", " << frame.rows << "]" << std::endl;

            imshow("读取视频", frame); // 显示当前帧
            waitKey(50);
        }
    }
    waitKey();

    return 0;
}

File: capture.py
import numpy as np
import cv2 as cv

def __gstreamer_pipeline(
        camera_id,
        capture_width=1920,
        capture_height=1080,
        display_width=1920,
        display_height=1080,
        framerate=30,
        flip_method=0,
    ):
    return (
            "nvarguscamerasrc sensor-id=%d ! "
            "video/x-raw(memory:NVMM), "
            "width=(int)%d, height=(int)%d, "
            "format=(string)NV12, framerate=(fraction)%d/1 ! "
            "nvvidconv flip-method=%d ! "
            "video/x-raw, width=(int)%d, height=(int)%d, format=(string)BGRx ! "
            "videoconvert ! "
            "video/x-raw, format=(string)BGR ! appsink max-buffers=1 drop=True"
            % (
                    camera_id,
                    capture_width,
                    capture_height,
                    framerate,
                    flip_method,
                    display_width,
                    display_height,
            )
    )
   
stream = cv2.VideoCapture(__gstreamer_pipeline(camera_id=0, flip_method=2), cv2.CAP_GSTREAMER)